File: POH7/SantaFuku/Main.scala
/* Try POH
 * author: Leonardone @ NEETSDKASU
 */
object Main extends App {
    
    def das(da :List[List[Int]], xs :List[Int], ys:List[Int]) : (List[Int], List[Int]) = {
        da match {
        case List() =>
            (xs.sorted, ys.sorted)
        case h :: t =>
            val d :: a :: _ = h
            if (d == 0)
                das(t, a :: xs, ys)
            else
                das(t, xs, a :: ys)
        }
    }
    
    def gis() :List[Int] = scala.io.StdIn.readLine.split(' ').map((x :String) => x.toInt).toList
    val x :: y :: z :: n :: _ = gis()
    val (xs, ys) = das((1 to n).map((_) => gis()).toList, List(0,x), List(0,y))
    val f = (ab :(Int, Int)) => { val (a, b) = ab; a - b }
    val xmin = xs.tail.zip(xs).map(f).min
    val ymin = ys.tail.zip(ys).map(f).min
    val ans = xmin * ymin * z
    println(ans)
}

File: POH7/SantaFuku/Main.hs
-- Try POH
-- author: Leonardone @ NEETSDKASU

import qualified Data.List as List (sort)

ti s = read s ::Int
gs = getLine
gi = fmap ti gs
gss = fmap words gs
gis = fmap (map ti) gss
ngt n f = sequence [f | _ <- [1..n]]
ngs n = ngt n gs
ngi n = ngt n gi
ngss n = ngt n gss
ngis n = ngt n gis

main = do
    (x:y:z:n:_) <- gis
    da <- ngis n
    print $ solve x y z n da

solve x y z n da = ans where
    append (d:a:_) (xs, ys)
        | d == 0    = (a:xs, ys)
        | otherwise = (xs, a:ys)
    (xs, ys) = foldr append ([0,x],[0,y]) da
    addval (a, b) = b - a
    valmin vs = let vss = List.sort vs in minimum . map addval $ zip vss $ tail vss
    xmin = valmin xs
    ymin = valmin ys
    ans = xmin * ymin * z
